### PR TITLE
[Benchmarks] Add a whole lot more benchmarks for Data

### PR DESCRIPTION
**Pull request by phausler**
This adds a number of new benchmarks for Data that test specific common cases of empty, small, medium and large Datas. Additionally this adds new benchmarks for creation as well as conversion from String to Data using Sequence/Collection initializers.

### CODE REVIEW

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
@swift-ci please smoke benchmark

File: benchmark/single-source/DataBenchmarks.swift [hunk 1]
```diff
@@ -414,3 +547,104 @@ public func run_AppendDataLargeToLarge(_ N: Int) {
...
+    }
+}
+
+@inline(never)
+public func run_createMediumArray(_ N: Int) {
+    for _ in 0..<10000 * N {
+        blackHole(Data([0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6, 0, 1, 2, 3, 4, 5, 6]))
+    }
+}
+
```
Comment (phausler) [inline benchmark/single-source/DataBenchmarks.swift]: @milseman What other converters do we want to test here? Should we test the Foundation overlay variations as well?

**Review by itaiferber — APPROVED** [review]
LGTM — as long as CI is happy, we should merge.

File: benchmark/single-source/DataBenchmarks.swift (hunk 1 above)
Comment (milseman) [reply benchmark/single-source/DataBenchmarks.swift]: Up to you. Next release, we'll want to unify them as part of some upcoming string spring cleaning.

**Comment by palimondo** [comment]
Among other things, this has also reintroduced setup overhead to some of the benchmarks I [previously cleaned up](https://github.com/phausler/swift/commit/ae9f5f18b02dda11ce07099fb8a107ee43d64d36#diff-a0f9e99147de6a460ae8c7484baef119). All the new `Large` variants are ~100% overhead of buffer creation. I see they were disabled in #20411, but I really don't think benchmarks that allocate over 1GB of memory are something that belongs to Swift Benchmarking Suite, at all. 

Also all the `DataCreate` measures is the performance of `arc4random_buf`. I don't see why we need truly random data for these tests… pseudorandom would be just as good, but since we are not doing anything with it, why not just fill it with repeated sequence of bytes like we do in [`IterateData`](https://github.com/apple/swift/blob/3ff92efdac709f400ca03b6b93136953d1300123/benchmark/single-source/IterateData.swift#L22-L31)?

I'll clean this up, but in the future, please do run `Benchmark_Driver check` locally before committing modified benchmarks. We now have hooked this up to CI, but it only validates newly added benchmarks. Here's what it say on my 2008 MBP:

<details>
<summary>Benchmark Check Report</summary>

✅  | Benchmark Check Report
---|---
⛔️⏱ | `DataAccessBytesLarge` has setup overhead of 15980062 μs (102.0%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⚠️Ⓜ️ | `DataAccessBytesLarge` has very wide range of memory used between independent, repeated measurements. <br><sub> `DataAccessBytesLarge` mem_pages [i1, i2]: min=[258239, 180882] 𝚫=77357 R=[3927, 81283]
⛔️⏱ | `DataAccessBytesMedium` has setup overhead of 16 μs (7.4%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataAppendArray` execution took at least 16619 μs. <br><sub> Decrease the workload of `DataAppendArray` by a factor of 32 (100), to be less than 1000 μs.
⛔️⏱ | `DataAppendBytesLarge` has setup overhead of 14701060 μs (92.3%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataAppendBytesLarge` execution took at least 1231514 μs (excluding the setup overhead). <br><sub> Decrease the workload of `DataAppendBytesLarge` by a factor of 2048 (10000), to be less than 1000 μs.
⛔️⏱ | `DataAppendBytesMedium` execution took at least 15679 μs. <br><sub> Decrease the workload of `DataAppendBytesMedium` by a factor of 16 (100), to be less than 1000 μs.
⚠️🔤 | `DataAppendDataLargeToLarge` name is composed of 6 words. <br><sub> Split `DataAppendDataLargeToLarge` name into dot-separated groups and variants. See http://bit.ly/BenchmarkNaming
⛔️⏱ | `DataAppendDataLargeToLarge` has setup overhead of 27390 μs (13.1%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataAppendDataLargeToLarge` execution took at least 181985 μs (excluding the setup overhead). <br><sub> Decrease the workload of `DataAppendDataLargeToLarge` by a factor of 256 (1000), to be less than 1000 μs.
⚠️Ⓜ️ | `DataAppendDataLargeToLarge` has very wide range of memory used between independent, repeated measurements. <br><sub> `DataAppendDataLargeToLarge` mem_pages [i1, i2]: min=[119, 95] 𝚫=24 R=[18, 29]
⚠️🔤 | `DataAppendDataLargeToMedium` name is composed of 6 words. <br><sub> Split `DataAppendDataLargeToMedium` name into dot-separated groups and variants. See http://bit.ly/BenchmarkNaming
⛔️⏱ | `DataAppendDataLargeToMedium` execution took at least 80878 μs. <br><sub> Decrease the workload of `DataAppendDataLargeToMedium` by a factor of 128 (100), to be less than 1000 μs.
⚠️Ⓜ️ | `DataAppendDataLargeToMedium` has very wide range of memory used between independent, repeated measurements. <br><sub> `DataAppendDataLargeToMedium` mem_pages [i1, i2]: min=[54, 51] 𝚫=3 R=[19, 22]
⚠️🔤 | `DataAppendDataLargeToSmall` name is composed of 6 words. <br><sub> Split `DataAppendDataLargeToSmall` name into dot-separated groups and variants. See http://bit.ly/BenchmarkNaming
⛔️⏱ | `DataAppendDataLargeToSmall` execution took at least 76184 μs. <br><sub> Decrease the workload of `DataAppendDataLargeToSmall` by a factor of 128 (100), to be less than 1000 μs.
⚠️🔤 | `DataAppendDataMediumToLarge` name is composed of 6 words. <br><sub> Split `DataAppendDataMediumToLarge` name into dot-separated groups and variants. See http://bit.ly/BenchmarkNaming
⛔️⏱ | `DataAppendDataMediumToLarge` has setup overhead of 19424 μs (17.8%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataAppendDataMediumToLarge` execution took at least 89827 μs (excluding the setup overhead). <br><sub> Decrease the workload of `DataAppendDataMediumToLarge` by a factor of 128 (100), to be less than 1000 μs.
⚠️Ⓜ️ | `DataAppendDataMediumToLarge` has very wide range of memory used between independent, repeated measurements. <br><sub> `DataAppendDataMediumToLarge` mem_pages [i1, i2]: min=[52, 53] 𝚫=1 R=[24, 23]
⚠️🔤 | `DataAppendDataMediumToMedium` name is composed of 6 words. <br><sub> Split `DataAppendDataMediumToMedium` name into dot-separated groups and variants. See http://bit.ly/BenchmarkNaming
⛔️⏱ | `DataAppendDataMediumToMedium` execution took at least 14957 μs. <br><sub> Decrease the workload of `DataAppendDataMediumToMedium` by a factor of 16 (100), to be less than 1000 μs.
⚠️🔤 | `DataAppendDataMediumToSmall` name is composed of 6 words. <br><sub> Split `DataAppendDataMediumToSmall` name into dot-separated groups and variants. See http://bit.ly/BenchmarkNaming
⛔️⏱ | `DataAppendDataMediumToSmall` execution took at least 14235 μs. <br><sub> Decrease the workload of `DataAppendDataMediumToSmall` by a factor of 16 (100), to be less than 1000 μs.
⚠️🔤 | `DataAppendDataSmallToLarge` name is composed of 6 words. <br><sub> Split `DataAppendDataSmallToLarge` name into dot-separated groups and variants. See http://bit.ly/BenchmarkNaming
⛔️⏱ | `DataAppendDataSmallToLarge` execution took at least 120114 μs. <br><sub> Decrease the workload of `DataAppendDataSmallToLarge` by a factor of 128 (1000), to be less than 1000 μs.
⚠️Ⓜ️ | `DataAppendDataSmallToLarge` has very wide range of memory used between independent, repeated measurements. <br><sub> `DataAppendDataSmallToLarge` mem_pages [i1, i2]: min=[74, 61] 𝚫=13 R=[12, 18]
⚠️🔤 | `DataAppendDataSmallToMedium` name is composed of 6 words. <br><sub> Split `DataAppendDataSmallToMedium` name into dot-separated groups and variants. See http://bit.ly/BenchmarkNaming
⛔️⏱ | `DataAppendDataSmallToMedium` execution took at least 13885 μs. <br><sub> Decrease the workload of `DataAppendDataSmallToMedium` by a factor of 16 (100), to be less than 1000 μs.
⚠️🔤 | `DataAppendDataSmallToSmall` name is composed of 6 words. <br><sub> Split `DataAppendDataSmallToSmall` name into dot-separated groups and variants. See http://bit.ly/BenchmarkNaming
⛔️⏱ | `DataAppendDataSmallToSmall` has setup overhead of 1210 μs (9.0%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataAppendDataSmallToSmall` execution took at least 12171 μs (excluding the setup overhead). <br><sub> Decrease the workload of `DataAppendDataSmallToSmall` by a factor of 16 (100), to be less than 1000 μs.
⛔️⏱ | `DataAppendSequence` execution took at least 65873 μs. <br><sub> Decrease the workload of `DataAppendSequence` by a factor of 128 (100), to be less than 1000 μs.
⛔️⏱ | `DataCopyBytesLarge` execution took at least 81330 μs. <br><sub> Decrease the workload of `DataCopyBytesLarge` by a factor of 128 (100), to be less than 1000 μs.
⚠️Ⓜ️ | `DataCopyBytesLarge` has very wide range of memory used between independent, repeated measurements. <br><sub> `DataCopyBytesLarge` mem_pages [i1, i2]: min=[34, 34] 𝚫=0 R=[13, 24]
⚠️⏱ | `DataCopyBytesMedium` execution took at least 1352 μs. <br><sub> Decrease the workload of `DataCopyBytesMedium` by a factor of 2 (10), to be less than 1000 μs.
⛔️⏱ | `DataCountLarge` has setup overhead of 15325398 μs (99.0%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataCountLarge` execution took at least 160287 μs (excluding the setup overhead). <br><sub> Decrease the workload of `DataCountLarge` by a factor of 256 (1000), to be less than 1000 μs.
⛔️⏱ | `DataCountMedium` has setup overhead of 18 μs (22.8%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataCountSmall` has setup overhead of 4 μs (7.5%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataCreateEmptyArray` execution took at least 14273 μs. <br><sub> Decrease the workload of `DataCreateEmptyArray` by a factor of 16 (100), to be less than 1000 μs.
⛔️⏱ | `DataCreateLarge` has setup overhead of 14540010 μs (95.3%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataCreateLarge` execution took at least 714406 μs (excluding the setup overhead). <br><sub> Decrease the workload of `DataCreateLarge` by a factor of 1024 (1000), to be less than 1000 μs.
⛔️⏱ | `DataCreateMedium` execution took at least 157904 μs. <br><sub> Decrease the workload of `DataCreateMedium` by a factor of 256 (1000), to be less than 1000 μs.
⚠️⏱ | `DataCreateMediumArray` execution took at least 9136 μs. <br><sub> Decrease the workload of `DataCreateMediumArray` by a factor of 16 (10), to be less than 1000 μs.
⛔️⏱ | `DataCreateSmall` execution took at least 223690 μs. <br><sub> Decrease the workload of `DataCreateSmall` by a factor of 256 (1000), to be less than 1000 μs.
⛔️⏱ | `DataCreateSmallArray` execution took at least 16932 μs. <br><sub> Decrease the workload of `DataCreateSmallArray` by a factor of 32 (100), to be less than 1000 μs.
⛔️⏱ | `DataMutateBytesLarge` has setup overhead of 18031434 μs (100.5%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataMutateBytesMedium` execution took at least 12779 μs. <br><sub> Decrease the workload of `DataMutateBytesMedium` by a factor of 16 (100), to be less than 1000 μs.
⚠️⏱ | `DataMutateBytesSmall` execution took at least 3367 μs. <br><sub> Decrease the workload of `DataMutateBytesSmall` by a factor of 4 (10), to be less than 1000 μs.
⛔️⏱ | `DataReplaceLarge` execution took at least 84975 μs. <br><sub> Decrease the workload of `DataReplaceLarge` by a factor of 128 (100), to be less than 1000 μs.
⛔️⏱ | `DataReplaceLargeBuffer` execution took at least 186047 μs. <br><sub> Decrease the workload of `DataReplaceLargeBuffer` by a factor of 256 (1000), to be less than 1000 μs.
⚠️Ⓜ️ | `DataReplaceLargeBuffer` has very wide range of memory used between independent, repeated measurements. <br><sub> `DataReplaceLargeBuffer` mem_pages [i1, i2]: min=[68, 61] 𝚫=7 R=[13, 26]
⛔️⏱ | `DataReplaceMedium` execution took at least 16692 μs. <br><sub> Decrease the workload of `DataReplaceMedium` by a factor of 32 (100), to be less than 1000 μs.
⛔️⏱ | `DataReplaceMediumBuffer` execution took at least 33516 μs. <br><sub> Decrease the workload of `DataReplaceMediumBuffer` by a factor of 64 (100), to be less than 1000 μs.
⛔️⏱ | `DataReplaceSmall` execution took at least 11302 μs. <br><sub> Decrease the workload of `DataReplaceSmall` by a factor of 16 (100), to be less than 1000 μs.
⛔️⏱ | `DataReplaceSmallBuffer` execution took at least 27040 μs. <br><sub> Decrease the workload of `DataReplaceSmallBuffer` by a factor of 32 (100), to be less than 1000 μs.
⛔️⏱ | `DataReset` execution took at least 10305 μs. <br><sub> Decrease the workload of `DataReset` by a factor of 16 (100), to be less than 1000 μs.
⛔️⏱ | `DataSetCountLarge` has setup overhead of 498 μs (17.3%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⚠️⏱ | `DataSetCountLarge` execution took at least 2377 μs (excluding the setup overhead). <br><sub> Decrease the workload of `DataSetCountLarge` by a factor of 4 (10), to be less than 1000 μs.
⚠️⏱ | `DataSetCountMedium` execution took at least 2477 μs. <br><sub> Decrease the workload of `DataSetCountMedium` by a factor of 4 (10), to be less than 1000 μs.
⛔️⏱ | `DataSubscriptLarge` has setup overhead of 14769664 μs (97.1%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataSubscriptLarge` execution took at least 443749 μs (excluding the setup overhead). <br><sub> Decrease the workload of `DataSubscriptLarge` by a factor of 512 (1000), to be less than 1000 μs.
⛔️⏱ | `DataSubscriptMedium` has setup overhead of 16 μs (10.0%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.
⛔️⏱ | `DataToStringMedium` has setup overhead of 44 μs (5.8%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`.

<details>

**Comment by phausler** [comment]
Large variants are useful in the regards that it takes a completely different code path and has different performance characteristics. I understand we should not do that for every test but it still should not regress imho.

The arc4random is being used to ensure the kernel does not just hand out a non faulting page of memory. Just being an allocation is not enough to properly identify used cases. It must be written to: and patterns like repeated are a completely different code path in the sequence initialization.

**Comment by phausler** [comment]
If we want to transition to a different random source that is fine in my book. It does not need to be arc4 but it should be something portable to Linux is all.

**Comment by palimondo** [comment]
Can you clarify this part?
> It must be written to: and patterns like repeated are a completely different code path in the sequence initialization.

To the best of my understanding, even the `.veryLarge` case just calls `sampleData(size: 1024 * 1024 * 1024 + 128)` which looks like this:

````swift
func sampleData(size: Int) -> Data {
    var data = Data(count: size)
    data.withUnsafeMutableBytes { getRandomBuf(baseAddress: $0, count: size) }
    return data
}
````

How is it different initialization pattern than this from IterateData?

````swift
  var data = Data(count: 16 * 1024)
  let n = data.count
  data.withUnsafeMutableBytes { (ptr: UnsafeMutablePointer<UInt8>) -> () in
    for i in 0..<n {
      ptr[i] = UInt8(i % 23)
    }
  }
````

I think that from the perspective of `Data`, which hands an `UnsafeMutablePointer` to the closure that fills it to capacity. As long as all the content is written to, it should look the same from kernel's perspective, too. Why do we need random data here?

**Comment by phausler** [comment]
Oh that differential is fine, I was just saying that Repeated (the sequence type) would behave differently.

**Pull request merged**